python/ray/serve/controller.py

```diff
@@ -300,32 +283,95 @@ def _scale_backend_replicas(self, backends: Dict[BackendTag, BackendInfo],
-                # pending tasks still executing on the replica. However, if we
-                # use replica.__ray_terminate__, we may send it while the
-                # replica is being restarted and there's no way to tell if it
-                # successfully killed the worker or not.
-                ray.kill(replica, no_restart=True)
 
-        self.backend_replicas_to_stop.clear()
+                async def kill_actor(replica_name_to_use):
+                    # NOTE: the replicas may already be stopped if we failed
+                    # after stopping them but before writing a checkpoint.
+                    try:
+                        replica = ray.get_actor(replica_name_to_use)
+                    except ValueError:
+                        return
+
+                    # TODO(edoakes): this logic isn't ideal because there may
+                    # be pending tasks still executing on the replica. However,
+                    # if we use replica.__ray_terminate__, we may send it while
+                    # the replica is being restarted and there's no way to tell
+                    # if it successfully killed the worker or not.
+                    ray.kill(replica, no_restart=True)
+
+                self.currently_stopping_replicas[asyncio.ensure_future(
+                    kill_actor(replica_name))] = tuple(
+                        [backend_tag, replica_tag])
+
+    async def _backend_control_loop(self):
+        start = time.time()
+        prev_warning = start
+        while self.currently_starting_replicas or \
+                self.currently_stopping_replicas:
+            if time.time() - prev_warning > REPLICA_STARTUP_TIME_WARNING_S:
+                prev_warning = time.time()
+                logger.warning("Waited {:.2f}s for replicas to start up. Make "
+                               "sure there are enough resources to create the "
+                               "replicas.".format(time.time() - start))
+
+            if self.currently_starting_replicas:
+                done, _ = await asyncio.wait(
+                    [
+                        x.as_future()
+                        for x in self.currently_starting_replicas.keys()
+                    ],
+                    timeout=0)
+                for fut in done:
+                    (backend_tag, replica_tag,
+                     replica_handle) = self.currently_starting_replicas.pop(
+                         fut.object_ref)
+                    self.backend_replicas[backend_tag][
+                        replica_tag] = replica_handle
+                    try:
+                        self.backend_replicas_to_start.get(backend_tag).remove(
+                            replica_tag)
+                        if len(self.backend_replicas_to_start[
+                                backend_tag]) == 0:
+                            del self.backend_replicas_to_start[backend_tag]
+                    except Exception:
+                        pass
```

Comment:
why do we need this? blinding ignoring exceptions looks dangerous

Reply:
I switched to more carefully checking before removing.